src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+
+#include "util/irep_ids.h"
+
+#include <algorithm>
+#include <regex>
+
+#ifdef JAVA_NONDET_DEBUG
+#include <iostream>
+#endif
+
+/*******************************************************************\
+
+  Struct: nondet_instruction_infot
+
+ Purpose: Hold information about a nondet instruction invocation.
+
+\*******************************************************************/
+
+struct nondet_instruction_infot final
+{
+  bool is_nondet_instruction;
+  bool is_never_null;
+};
+
+/*******************************************************************\
+
+Function: is_nondet_returning_object
+
+  Inputs:
+    function_call: A function call expression.
+
+ Outputs: Whether the function has a nondet signature, and whether the nondet
+          item may be null.
+
+
+ Purpose: Find whether the supplied function call has a recognised 'nondet'
+          library signature.  If so, specify whether the function call is
+          expected to return non-null.
+
+\*******************************************************************/
+
+static nondet_instruction_infot is_nondet_returning_object(
+  const code_function_callt &function_call)
+{
+  const auto function_name=
+    id2string(function_call.function().get(ID_identifier));
+  std::smatch match_results;
+  if(!std::regex_match(
+       function_name,
+       match_results,
+       std::regex(".*org\\.cprover\\.CProver\\.nondetWith(out)?Null.*")))
+  {
+    return {false, false};
+  }
+
+  return {true, match_results[1].matched};
+}
+
+/*******************************************************************\
+
+Function: get_nondet_instruction_info
+
+  Inputs:
+    instr: A goto program instruction.
+
+ Outputs: Whether the instruction is a function with a nondet signautre, and
+          whether the nondet item may be null.
+
+ Purpose: Return whether the instruction has a recognised 'nondet' library
+          signature.
+
+\*******************************************************************/
+
+static nondet_instruction_infot get_nondet_instruction_info(
+  const goto_programt::targett &instr)
+{
+  if(!(instr->is_function_call() && instr->code.id()==ID_code))
+  {
+    return {false, false};
+  }
+  const auto &code=to_code(instr->code);
+  if(code.get(ID_statement)!=ID_function_call)
+  {
+    return {false, false};
+  }
+  const auto &function_call=to_code_function_call(code);
+  return is_nondet_returning_object(function_call);
+}
+
+// TODO Pascal suggested moving the following function to util, but I'm not
+// sure.  If goto_programt has public data members, this implies that any state
+// that these data members takes is valid for the goto_program as a whole,
+// which is obviously not the case - if I insert a node into
+// goto_programt::instructions without updating the iterator positions, this
+// has a good chance of invalidating the program.  It should be *impossible* to
+// place an object (with the strong exception safety guarantee) into an invalid
```

Comment:
Yes; this is a known issue and is documented elsewhere.  Fixing this is an entirely different task.  There are not only structural but semantic issues here.  When I / someone in QA has a chance, we need a "check well formed" method for goto_functiont that at least documents some of the invariants that must hold on the structure.